src/query/functions/temporal/functions.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package temporal
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/m3db/m3/src/query/executor/transform"
+)
+
+const (
+	// ResetsType returns the number of counter resets within the provided time range as a time series.
+	// Any decrease in the value between two consecutive datapoints is interpreted as a counter reset.
+	// ResetsTemporalType should only be used with counters.
+	ResetsType = "resets"
+
+	// ChangesType returns the number of times a value changes within the provided time range for
+	// a given time series.
+	ChangesType = "changes"
+)
+
+var (
+	temporalFuncs = map[string]interface{}{
+		ResetsType:  struct{}{},
+		ChangesType: struct{}{},
+	}
+)
+
+// NewFunctionOp creates a new base temporal transform for functions
+func NewFunctionOp(args []interface{}, optype string) (transform.Params, error) {
+	if _, ok := temporalFuncs[optype]; ok {
```

Comment:
Rather than a map lookup here, create the compFunc here and pass it into your function node

Reply:
Hmm then we'd need to change the definition of `newBaseOp`, i think this way is cleaner

Reply:
These functions, aggregation, and irate all look very similar. Can you refactor them into a single path rather than have a bunch of different ops created instead? The only thing they differ in is what the function in `Process()` looks like

Reply:
I'll do this in a separate pr